experimental/InvariantTheory/iterators.jl

```diff
@@ -0,0 +1,313 @@
+################################################################################
+#
+#  All Monomials
+#
+################################################################################
+
+# Returns an iterator over all monomials of R of degree d
+all_monomials(R::MPolyRing, d::Int) = AllMonomials(R, d)
+
+AllMonomials(R::MPolyRing, d::Int) = AllMonomials{typeof(R)}(R, d)
+
+Base.eltype(AM::AllMonomials) = elem_type(AM.R)
+
+# We are basically doing d-multicombinations of n here and those are "the same"
+# as d-combinations of n + d - 1 (according to Knuth).
+Base.length(AM::AllMonomials) = binomial(nvars(AM.R) + AM.d - 1, AM.d)
+
+function Base.iterate(AM::AllMonomials, state::Nothing = nothing)
+  n = nvars(AM.R)
+  if AM.d == 0
+    s = zeros(Int, n)
+    s[n] = AM.d + 1
+    return one(AM.R), s
+  end
+
+  c = zeros(Int, n)
+  c[1] = AM.d
+  s = zeros(Int, n)
+  if isone(n)
+    s[1] = AM.d + 1
+    return (set_exponent_vector!(one(AM.R), 1, c), s)
+  end
+
+  s[1] = AM.d - 1
+  s[2] = 1
+  return (set_exponent_vector!(one(AM.R), 1, c), s)
+end
+
+function Base.iterate(AM::AllMonomials, s::Vector{Int})
+  n = nvars(AM.R)
+  d = AM.d
+  if s[n] == d + 1
+    return nothing
+  end
+  c = copy(s)
+  if s[n] == d
+    s[n] += 1
+    return (set_exponent_vector!(one(AM.R), 1, c), s)
+  end
+
+  for i = n - 1:-1:1
+    if !iszero(s[i])
+      s[i] -= 1
+      if i + 1 == n
+        s[n] += 1
+      else
+        s[i + 1] = 1
+        if !iszero(s[n])
+          s[i + 1] += s[n]
+          s[n] = 0
+        end
+      end
+      return (set_exponent_vector!(one(AM.R), 1, c), s)
+    end
+  end
+end
+
+function Base.show(io::IO, AM::AllMonomials)
+  print(io, "Iterator over the monomials of degree $(AM.d) of\n")
+  print(io, AM.R)
+end
+
+################################################################################
+#
+#  Bases of Invariant Rings
+#
+################################################################################
+
+# Returns the dimension of the graded component of degree d.
+# If we cannot compute the Molien series (so far in the modular case), we return
+# -1.
```

Comment:
Why don't you throw an exception instead?

Reply:
I don't really want the function to fail. If it is possible to compute the dimension via the Molien series, I want to have it, otherwise I have to make do without it.
I can also let it return a `Bool` and an `Int`, so that the first output says whether the second one makes sense.
This is not supposed to be a "user-function" in general.